### PR TITLE
[tcling] Use more variables to denote clearly the state and intent.

### PR DESCRIPTION
**Pull request by vgvassilev**
We might be in a situation where we linked the library into an executable and
our trigger function is inside it. There is no point in trying to load rdict
and pcm files as it will fail. In that case we exit early avoiding the
invalid state pointed by an assert.

This patch is another try to fix the assert in the incrementals.

@Axel-Naumann, I have tested and this patch fixes the assert. However it does not fix the test failure. I checked with a local alternative patch which did not have the early exit but just avoided the LoadPCM section and the test fails in the same way.

### CODE REVIEW

File: core/metacling/src/TCling.cxx [hunk 1]
```diff
@@ -1783,6 +1783,14 @@ void TCling::RegisterModule(const char* modulename,
       code += payloadCode;
 
    const char* dyLibName = nullptr;
+   // const char* dyLibName = FindLibraryName(triggerFunc);
+
+   // The triggerFunc may not be in a shared object but in an executable.
+   // In that case, we are done.
+   // bool isSharedLib = dyLibName && cling::DynamicLibraryManager::isSharedLibrary(dyLibName);
```
Comment (Axel-Naumann) [inline core/metacling/src/TCling.cxx]: What does that commented code line mean, is this part of the doc or a left-over?

File: core/metacling/src/TCling.cxx [hunk 2]
```diff
@@ -1798,9 +1806,12 @@ void TCling::RegisterModule(const char* modulename,
       if (dyLibName && !cling::DynamicLibraryManager::isSharedLibrary(dyLibName))
          dyLibName = nullptr;
 
-      if (dyLibName) {
 
-         // We were able to determine that is a valid library.
+      if (dyLibName) {
+         // We need to open the dictionary shared library, to resolve symbols
```
Comment (Axel-Naumann) [inline core/metacling/src/TCling.cxx]: Why do we have the same comment here as in line 1797, in the new version of the file?

File: core/metacling/src/TCling.cxx [hunk 3]
```diff
@@ -1964,7 +1976,15 @@ void TCling::RegisterModule(const char* modulename,
       }
    }
 
-   if (gIgnoredPCMNames.find(modulename) == gIgnoredPCMNames.end()) {
+   // This is an executable. There is no point in trying to load rdict and pcm
```
Comment (Axel-Naumann) [inline core/metacling/src/TCling.cxx]: The "virtual readers' instruction pointer" can arrive at the comment even if the library is a shared library and not an executable. Maybe reword the comment: "If the library was not dlopened it's likely an executable. There is no point..."

File: core/metacling/src/TCling.cxx [hunk 4]
```diff
@@ -1964,7 +1976,15 @@ void TCling::RegisterModule(const char* modulename,
...
+   // files.
+   // FIXME: This means that we have linked the `modulename` in an executable
+   // where we resolve our trigger function. It is an open question whether we
+   // should try loading the `modulename`[_rdict].pcms.
+   // if (!isSharedLib)
```
Comment (Axel-Naumann) [inline core/metacling/src/TCling.cxx]: Why is this commented code left here?

File: core/metacling/src/TCling.cxx [hunk 5]
```diff
@@ -1964,7 +1976,15 @@ void TCling::RegisterModule(const char* modulename,
...
+   //    return;
+
+   if (wasDlopened && gIgnoredPCMNames.find(modulename) == gIgnoredPCMNames.end()) {
       llvm::SmallString<256> pcmFileNameFullPath;
       if (dyLibName) {
```
Comment (Axel-Naumann) [inline core/metacling/src/TCling.cxx]: Doesn't `wasDlopened` now imply `dylibName`?

File: core/metacling/src/TCling.cxx (hunk 1 above)
Comment (vgvassilev) [reply core/metacling/src/TCling.cxx]: This is a WIP, trial and error development. This makes sure I remember what was done before...

**Comment by oshadura** [comment]
@vgvassilev I will take a look on IO failures :)

File: core/metacling/src/TCling.cxx (hunk 1 above)
Comment (Axel-Naumann) [reply core/metacling/src/TCling.cxx]: OK please label this as "do not review yet" or change the title to include [WIP] to let us know about the stage...

**Comment by vgvassilev** [comment]
@oshadura, I think I found the bug, I will ping you if I need help.

**Comment by vgvassilev** [comment]
@phsft-bot build with flags -DCMAKE_BUILD_TYPE=Debug -DLLVM_BUILD_TYPE=Debug

**Comment by vgvassilev** [comment]
The cmsUnload test fails on other PRs, too.

**Comment by Axel-Naumann** [comment]
> The cmsUnload test fails on other PRs, too.

It fails in the `assert` that this PR claims to fix (but apparently doesn't).

**Comment by Axel-Naumann** [comment]
The failures of `unloadCMS` are due to the assert; the failures here are new.

**Comment by vgvassilev** [comment]
@phsft-bot build!

**Comment by Axel-Naumann** [comment]
@phsft-bot build!

**Comment by Axel-Naumann** [comment]
Windows failure is understood!
